File: LeetCodeSolutions/problems/3sum/solution.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        // Sort the Input Array to skip duplicates
        sort(nums.begin(), nums.end());
        
        vector<vector<int>> res;
        
        // Fix the first two numbers  
        // Search for the third number which equals to
        // z = -( x + y ) and append to the result
        
        for(int i = 0; i < nums.size(); i++){
            int lptr = i+1;
            int rptr = nums.size() - 1;
        
            // check for duplicate .. Exit loop and skip to the next 
            if( i > 0 && nums[i] == nums[i-1]){
                
                // Skip Duplicates 
                
            } else {
                
                while(lptr < rptr){
                    int summation = nums[i] + nums[lptr] + nums[rptr];                
                    
                    if(summation < 0){
                        lptr++;
                    }
                    else if(summation > 0){
                        rptr--;

                    }
                    else{
                        
                        res.push_back({nums[i], nums[rptr], nums[lptr]});
                        lptr++;
                        while(lptr < rptr && nums[lptr-1] == nums[lptr]){
                            lptr++;
                        }
                    }   
                }
            }
        }

        return res;
    }
};

File: LeetCodeSolutions/problems/3sum/solution.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:

        triplets = []
        nums.sort()

        for k in range(len(nums)):
            # Check for duplicate pointers
            if k > 0 and nums[k] == nums[k-1]:
                continue

            #  Two Pointers Approach
            i = k + 1
            j = len(nums) - 1
            
            while i < j:
                target = nums[i] + nums[k] + nums[j]

                if target < 0:
                    i += 1
                elif target > 0:
                    j -= 1
                else:
                    triplets.append([nums[i], nums[j], nums[k]])
                    i += 1
                    while nums[i] == nums[i-1] and i < j:
                        i += 1
        return triplets

